sensors/sensors.c

```diff
@@ -0,0 +1,600 @@
+	}
+
+	/* Assign position of events for each device and calculate number of events of all types */
+	for (node = lib_rbMinimum(sensors_common.infos.root), i = 0; node != NULL; node = lib_rbNext(node), ++i) {
+		info = lib_treeof(sensor_info_t, node, node);
+
+		for (j = 0; j < NB_SENSOR_TYPES; ++j) {
+			id = 1 << j;
+			if ((info->types & id) != 0) {
+				sensors_common.devEvents[i][j] = sensors_common.evtNb[j]++;
+			}
+		}
+	}
+
+	/* Allocate memory for events tables */
+	for (i = 0; i < NB_SENSOR_TYPES; ++i) {
+		if (sensors_common.evtNb[i] != 0) {
+			sensors_common.events[i] = malloc(sizeof(sensor_event_t) * sensors_common.evtNb[i]);
+			if (sensors_common.events[i] == NULL) {
+				fprintf(stderr, "sensors: cannot allocate memory for events\n");
+				continue;
+			}
+
+			res = mutexCreate(&sensors_common.locks[i]);
+			if (res < 0) {
+				free(sensors_common.events[i]);
+				fprintf(stderr, "sensors: cannot create lock\n");
+			}
+		}
+	}
+
+	return EOK;
+}
+
+
+static int sensors_drvInit(const char *name, const char *args)
+{
+	int res;
+	sensor_info_t *info;
+	const sensor_drv_t *drv;
+
+	drv = sensors_getDrv(name);
+	if (drv == NULL) {
+		fprintf(stderr, "sensors: cannot find driver for %s\n", name);
+		return -EINVAL;
+	}
+
+	info = malloc(sizeof(sensor_info_t));
+	if (info == NULL) {
+		fprintf(stderr, "sensors: cannot allocate memory\n");
+		return -ENOMEM;
+	}
+
+	res = drv->alloc(info, args);
+	if (res < 0) {
+		free(info);
+		fprintf(stderr, "sensors: cannot allocate sensor %s, err: %d\n", name, res);
+		return res;
+	}
+
+	res = idtree_alloc(&sensors_common.infos, &info->node);
+	if (res < 0) {
+		free(info);
+		fprintf(stderr, "sensors: cannot add sensor %s, to tree - err: %d\n", name, res);
+		return res;
+	}
+
+	info->id = res;
+	info->drv = drv->name;
+
+	return res;
+}
+
+
+static int sensors_parseArgs(int argc, char **argv)
+{
+	int c, devsz = 0;
+	char *args;
+	const char *name;
+
+	while ((c = getopt(argc, argv, "s:h")) != -1) {
+		switch (c) {
+			case 's': /* <sensor:args> */
+				name = optarg;
+				args = strchr(optarg, ':');
+				if (args == NULL) {
+					fprintf(stderr, "sensors: missing a sensors arguments: %s\n", optarg);
+					return devsz;
+				}
+
+				*args++ = '\0';
+				if (sensors_drvInit(name, args) >= 0) {
+					++devsz;
+				}
+				break;
+
+			case 'h':
+				sensors_help(argv[0]);
+				return 0;
+
+			default:
+				return -EINVAL;
+		}
+	}
+
+	return devsz;
+}
+
+
+static int sensors_createDev(void)
+{
+	oid_t root;
+	int err;
+
+	/* Wait for the filesystem */
+	while (lookup("/", NULL, &root) < 0) {
+		usleep(10000);
+	}
+
+	sensors_common.oid.id = 0;
+	err = portCreate(&sensors_common.oid.port);
+	if (err < 0) {
+		fprintf(stderr, "sensors: could not create a port\n");
+		return err;
+	}
+
+	err = create_dev(&sensors_common.oid, "/dev/sensors");
+	if (err) {
+		fprintf(stderr, "sensors: could not create a device\n");
+		return -EINVAL;
+	}
+
+	return EOK;
+}
+
+
+void sensors_cleanup(size_t devsz)
```

Comment:
I am not sure if this is really needed - we would rather want to deinit device drivers, memory management is not really needed - it's faster to leave this to the kernel. @nalajcie your opinion?